### DISCUSSION TITLE
How to actually use types from query builder?

### DISCUSSION THREAD

**Discussion by jakubmazanec** [geldata/gel#4467] [Q&A] [answered]
Generated query builder contains some useful interfaces. If I have this schema:

```
# dbschema/default.esdl
type User {
  required property name -> str;
  required property email -> str { constraint exclusive; };
}
```

Then the query builder contains:
```ts
// dbschema/edgeql-js/types.ts
export interface User extends std.$Object {
  email: string;
  name: string;
}
```

But `std.$Object` has required property `__type__` that complicates usage of `User`; if a need to pass user data, and I can't rely on type inference, I now cannot simply use `User` type:

```ts
import {type User} from '~/dbschema/edgeql-js'

function doStuffWithUser(user: User) {
 // do stuff
}

let user = await e.assert_single(e.select(e.User, (user) => ({
  ...e.User['*'],
  filter: // filter the user somehow
  limit: 1,
}))).run(db);

if (user) {
  doStuffWithUser(user); // Error here
}
```

I get error `Argument of type '{ id: string; name: string; email: string; }' is not assignable to parameter of type 'User'. Property '"__type__"' is missing in type '{ id: string; name: string; email: string; }' but required in type 'User'.`

Am I using the generated types in a wrong way? Or am I missing something else? Thanks for the help!

**Comment by jaclarke** [comment, marked as answer]
The generated interfaces reflect (as closely as possible) all the fields an object type has, along with their cardinality, eg. an optional property would be something like `name: string | null`, required property is `name: string`, multi is `name: string[]`. So since `__type__` is a required link on the base `std::Object`, that's what it gets reflected to. The idea being that you can then use `Partial`/`Required`/`Pick`/`Omit`/etc. to shape the interfaces to your needs, since you can select any arbitrary props/links of a type in a query. The discussion where we decided on this behaviour is here: https://github.com/edgedb/edgedb-js/pull/281.

**Reply by jaclarke** [reply]
We do need to improve the documentation on this though, currently generated interfaces are only briefly mentioned at the end of this page: https://www.edgedb.com/docs/clients/js/generation#generated-interfaces

**Reply by jakubmazanec** [reply]
Thanks for the reply!

Using `Pick`, `Omit`, etc. is obvious solution, but I because I can't image usecase for  `__type__`, I think the generated query builder should by default (or maybe this could be controlled via CLI arg?) export types already stripped of `__type__`, because this is basically what I do first for every object type before I use it elsewhere:

```ts
import {type SomeType as SomeTypeOriginal} from '~/dbschema/edgeql-js'

export type SomeType = Omit<SomeTypeOriginal, '__type__'>
```

**Reply by avedetvedea** [reply]
@colinhacks Would it be possible to export types without inheritance, to not to deal with the `Omit`s like above everywhere we need to pass the db entities around?

Currently, when I retrieve a db entity via the TS client, to use it as a payload value for a deserialize function (say for an authentication flow), the TS complains when the return type is a promise of the exported type. I expect these types to match, i.e., the type that the TS client gives me and the generated type from edgeql-js. 

My current workflow not only includes stripping out the '\__type__\' for all types whenever I want to pass db entities around (and possibly re-exporting them from another place to reduce bloat), but also includes things like conversions for enums. For instance, if I have a `level` type with `Level1`, `Level2`, `Level3` values, the TS client returns `Level1` | `Level2` | `Level3`, whereas the exported type exports an enum. Because both of the procedures are using the edgeql-js, I would expect the TS translation to be the same: they are both trying to convert some types from the same schema to the same language, after all. 

FWIW, exporting additional 'base types' would also be helpful, stripped out from the properties that are not edgedb or custom scalars (i.e., stripped out of its relations). All suggestions I made are from ORM workflows in TS space.
